Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp

```diff
@@ -717,5 +718,26 @@ bool TextUtil::hasPositionDependentContentWidth(StringView textContent)
     return charactersContain<UChar, tabCharacter>(textContent.span16());
 }
 
+
+char32_t TextUtil::baseCharacterFromGraphemeCluster(StringView graphemeCluster)
+{
+    for (auto codepoint : graphemeCluster.codePoints()) {
+        if (!isCombiningMark(codepoint))
+            return codepoint;
+    }
+    return graphemeCluster.characterAt(0);
+}
+
+char32_t TextUtil::lastBaseCharacterFromText(StringView string)
+{
+    StringView* lastCluster { nullptr };
+    auto graphemeClusters =  string.graphemeClusters();
+    for (auto cluster : graphemeClusters)
+        lastCluster = &cluster;
```

Comment:
Analyzing grapheme clusters for the entire string is quite expensive. Since we're just trying to find the last one, I'm pretty sure you can just walk backwards from the end of the string until you get to a non-combining-mark.

Reply:
That's a good point. The reason I did that is because `graphemeClusters` returns a class with `begin` and `end` iterators, so I could not skip directly to the last one. But as you mentioned, we are just interested in the last base character so instead of using grapheme cluster's api here we can walk back the string and find the first code point which is not a combining mark I guess.